rspace/src/main/scala/coop/rchain/rspace/RSpace.scala

```diff
@@ -223,107 +222,84 @@ class RSpace[F[_], C, P, A, K](
+              store.removeJoin(candidateChannel, channels)
+          }
+        }
+
+    def constructResult: MaybeActionResult =
+      Some(
+        (
+          ContResult[C, P, K](
             continuation,
             persistK,
-            peeks,
-            consumeRef
+            channels,
+            patterns,
+            peeks.nonEmpty
           ),
-          continuationIndex,
-          dataCandidates
-          ) =>
-        def registerCOMM: F[Unit] =
-          syncF.delay {
-            val produceRefs = dataCandidates.map(_.datum.source)
-            eventLog
-              .update(
-                COMM(consumeRef, produceRefs, peeks, produceCounters(produceRefs)) +: _
-              )
-          }
-
-        def maybePersistWaitingContinuation: F[Unit] =
-          if (!persistK) {
-            store.removeContinuation(
-              channels,
-              continuationIndex
-            )
-          } else ().pure[F]
-
-        def removeMatchedDatumAndJoin(channelsToIndex: Map[C, Int]): F[Seq[Unit]] =
-          dataCandidates
-            .sortBy(_.datumIndex)(Ordering[Int].reverse)
-            .traverse {
-              case DataCandidate(
-                  candidateChannel,
-                  Datum(_, persistData, _),
-                  _,
-                  dataIndex
-                  ) => {
-                (if (dataIndex >= 0 && !persistData) {
-                   store.removeDatum(candidateChannel, dataIndex)
-                 } else ().pure[F]) >> store.removeJoin(candidateChannel, channels)
-              }
-            }
-
-        def constructResult: MaybeActionResult =
-          Some(
-            (
-              ContResult[C, P, K](
-                continuation,
-                persistK,
-                channels,
-                patterns,
-                peeks.nonEmpty
-              ),
-              dataCandidates.map(
-                dc => Result(dc.channel, dc.datum.a, dc.removedDatum, dc.datum.persist)
-              )
-            )
+          dataCandidates.map(
+            dc => Result(dc.channel, dc.datum.a, dc.removedDatum, dc.datum.persist)
           )
+        )
+      )
 
-        for {
-          _               <- metricsF.incrementCounter(produceCommLabel)
-          _               <- registerCOMM
-          _               <- maybePersistWaitingContinuation
-          indexedChannels = channels.zipWithIndex.toMap
-          _               <- removeMatchedDatumAndJoin(indexedChannels)
-          _ <- logF.debug(
-                s"produce: matching continuation found at <channels: $channels>"
-              )
-        } yield constructResult
-    }
+    for {
+      _               <- metricsF.incrementCounter(produceCommLabel)
+      _               <- logComm(consumeRef, peeks, dataCandidates)
+      _               <- store.removeContinuation(channels, continuationIndex).unlessA(persistK)
+      indexedChannels = channels.zipWithIndex.toMap
```

Comment:
Here we're creating an intermediate collection (via `zipWithIndex`) for no good reason. It's generally better to use `.iterator` in cases like this